File: traffic_estimation/src/pav_algorithm/PAValgo.py
from numpy import array, inf, dot, ones, float

# PAV algorithm with box constraints
def pav_algo(y, w, l=-inf, u=inf):
	assert(y.size == w.size)
	n = len(y)
	x = y #x=y.copy()
	
	if n>1:
		j=range(n+1) # j contains the first index of each block
		ind = 0
		
		while ind < len(j)-2:
			if avg(y,w,j,ind+1) < avg(y,w,j,ind):
				j.pop(ind+1)
				while ind > 0 and avg(y,w,j,ind-1) > avg(y,w,j,ind):
					if avg(y,w,j,ind) <= avg(y,w,j,ind-1):
						j.pop(ind)
						ind -= 1
			else:
				ind += 1	
		
		#x = array([avg(y,w,j,i)*ones(j[i+1]-j[i]) for i in range(len(j)-1)])
		for i in range(len(j)-1): x[j[i]:j[i+1]] = avg(y,w,j,i)*ones(j[i+1]-j[i])
		
	return x

# weighted average
def avg(y,w,j,ind):
	block = range(j[ind],j[ind+1])
	#print block
	wB = w[block]
	return dot(wB,y[block])/float(wB.sum())

# DEMO starts here
if __name__ == "__main__":
	print """
Demonstration of the PAV algorithm on a small example."""
	print
	y = [4.,5.,1.,6.,8.,7.]
	w = array([1,1,1,1,1,1])
	y=array(map(float, y))
	print "y vector", y
	print "weights", w
	print "solution", pav_algo(y,w)
	

File: PAValgo.m
function x = PAValgo(y,w,l,u)

%{
Function that solves the IRC problem
    min. sum_i w_i(y_i-x_i)^2
    s.t. l <= x_1 <= ... <= x_n <= u
using PAV algorithm described in
Best and Chakravarti. Active Set Algorithms for isotonic regression; a unifying framework
%}

n = length(y);
x = y;

if n>1
    J = 1:(n+1); % J contains the first index of each block
    indB = 1;
    %n
    %length(w)
    while indB < length(J)-1
        %fprintf('Loop\n')
        %lenJ
        %length(J)
        %indB
        %n
        if avg(y,w,J,indB+1) < avg(y,w,J,indB)
            
            J(indB+1) = [];
            %lenJ = lenJ-1;
            
            while indB > 1 && avg(y,w,J,indB-1) > avg(y,w,J,indB)
                %fprintf('Inner Loop\n')
                if avg(y,w,J,indB) <= avg(y,w,J,indB-1)
                    J(indB) = [];
                    %lenJ = lenJ-1;
                    indB = indB-1;
                end
            end
            
        else
            
            indB = indB+1;
        end
        
    end
    %fprintf('Put x together\n')
    for i=1:length(J)-1
        ind = J(i):(J(i+1)-1);
        x(ind) = avg(y,w,J,i)*ones(length(ind),1);
    end
end

if l <= u
    x(x<l) = l;
    x(x>u) = u;
else
    fprintf('We must have l<=u, projection aborted!')
end

end

%=============================

function av = avg(y,w,J,indB)

%{
Function that computes average
%}
ind = J(indB):(J(indB+1)-1); % indices of B
%max(ind)
%length(w)
wB = w(ind);
av = wB'*y(ind) / sum(wB);

end